### ISSUE TITLE
How can I tell if ML.NET is using GPU acceleration?

### ISSUE THREAD

**Issue by CBrauer** [dotnet/machinelearning#5873]
Hey Guys,

I have a question about GPU support in ML.NET.

My environment is:
•	Visual Studio 2022 (Preview)
•	.NET 6.0 (Preview)

I have downloaded and installed “ML.NET Model Builder GPU Support” and tested the GPU with Model Builder.  Model Builder likes my GPU card as you can see:

![GPU Support](https://user-images.githubusercontent.com/1317234/124820113-ea22dc00-df21-11eb-911e-1e728c787066.png)

It works o.k.

I have some legacy ML.NET code, and I am wondering if I re-build my old code will it use the newly installed GPU drivers?

How can I tell if my ML.NET project uses the GPU?

Charles

**Comment by CBrauer** [comment]
I started binary classification model training program.
And then I ran the Task Manager to see what the GPU usage was:

![Gpu Usage](https://user-images.githubusercontent.com/1317234/124974936-42211780-dfe2-11eb-96bf-71db0f544c23.png)

My GPU usage was only 9%.  Why
Charles

**Comment by michaelgsharp** [comment]
@JakeRadMSFT Can you take a look here? I am not super familiar with how the model builder gpu stuff translates over to ML.NET.

**Comment by torronen** [comment]
@CBrauer This is probably image classification, right? If I understand correctly LightGBM does not support GPU. GPU is only used for getting embeddings from the images, so . At least, I could not see GPU support in any other tasks in Model Builder and it would be strange if LightGBM GPU is supported but it is not enabled in other tasks. My guess it is doing something like this https://docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/image-classification

So, maybe, GU is only used for loading the images and the LightGBM is ran to classify them on CPU so you should see also 100% CPU usage. BTW, I think, at least most of the time, GPU training in Tensorflow should show up in "3D" part of the GPU consumption.

On other hand, if LightGBM has GPU binary in Model builder, then I think it would be super-great to see it in all tasks.

**Comment by michaelgsharp** [comment]
I don't think that LightGBM has a GPU binary in Model Builder. In fact, I don't think that LightGBM has pre-compiled binaries for anything with a GPU, though you can compile it yourself.

It would be awesome if they did have some pre-compiled GPU binaries so we could use them. It should be possible for an end user to build it themselves and swap it out, though I haven't tested that myself.

**Comment by torronen** [comment]
I tried replacing Model builder's lightgbm dll binary compiled for Nvidia GPU in the extension folder. It works but uses CPU. I think it might require the device_type=GPU parameter. Just experimenting.
